apps/counter/app/tests/builder_tests.rs

```diff
@@ -0,0 +1,167 @@
+use essential_app_utils::compile::compile_pint_project;
+use regex::Regex;
+use std::process::Stdio;
+use tokio::{
+    io::{AsyncBufReadExt, BufReader},
+    process::{Child, Command as TokioCommand},
+    time::{sleep, Duration},
+};
+
+const PINT_DIRECTORY: &str = "../pint";
+
+#[tokio::test]
+async fn builder_integration() {
+    let (_builder_process, node_address, builder_address) = start_essential_builder().await;
+
+    let _ = compile_pint_project(concat!(env!("CARGO_MANIFEST_DIR"), "/../pint").into())
+        .await
+        .unwrap();
+    // ensure builder has started and compilation is done
+    sleep(Duration::from_secs(1)).await;
+
+    deploy_contract(builder_address.clone()).await;
+
+    let count = read_count(node_address.clone(), PINT_DIRECTORY).await;
+    assert_eq!(count, 0);
+
+    let returned_count = increment_count(
+        node_address.clone(),
+        builder_address.clone(),
+        PINT_DIRECTORY,
+    )
+    .await;
+    assert_eq!(returned_count, 1);
+
+    let new_count = read_count(node_address.clone(), PINT_DIRECTORY).await;
+    assert_eq!(new_count, returned_count);
+
+    let count = new_count;
+    let _ = increment_count(
+        node_address.clone(),
+        builder_address.clone(),
+        PINT_DIRECTORY,
+    )
+    .await;
+
+    let new_count = read_count(node_address.clone(), PINT_DIRECTORY).await;
+
+    assert_eq!(new_count, count + 1);
+}
+
+async fn read_count(node_address: String, pint_directory: &str) -> u32 {
+    let read_output = TokioCommand::new("cargo")
+        .args(&[
+            "run",
+            "--",
+            "read-count",
+            &format!("http://{}", node_address.as_str()),
+            pint_directory,
+        ])
+        .output()
+        .await
+        .expect("Failed to execute command");
+
+    assert!(read_output.status.success(), "Command failed to run");
+
+    let stdout_str = String::from_utf8(read_output.stdout).expect("Failed to parse stdout");
+
+    let _stderr_str = String::from_utf8(read_output.stderr).expect("Failed to parse stderr");
+
+    let count = stdout_str
+        .split_whitespace()
+        .last()
+        .expect("Failed to parse count")
+        .parse::<u32>()
+        .expect("Failed to parse count");
+    count
+}
+
+async fn increment_count(
+    node_address: String,
+    builder_address: String,
+    pint_directory: &str,
+) -> u32 {
+    let increment_output = TokioCommand::new("cargo")
+        .args(&[
+            "run",
+            "--",
+            "increment-count",
+            &format!("http://{}", node_address.as_str()),
+            &format!("http://{}", builder_address.as_str()),
+            pint_directory,
+        ])
```

Comment:
```suggestion
        .args([
            "run",
            "--",
            "increment-count",
            &format!("http://{}", node_address.as_str()),
            &format!("http://{}", builder_address.as_str()),
            pint_directory,
        ])
```
not sure why CI isn't picking it up but clippy is complaining locally about this.